rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,234 @@
+
+# Motivation
+[motivation]: #motivation
+
+- (Especially new) package contributors are having a hard time figuring out which files to add or edit. These are very common questions:
+  - Which directory should my package definition go in?
+  - What are all the categories and do they matter?
+  - What if the package has multiple matching categories?
+  - Why can't I build my package after adding the package file?
+  - Where in all-packages.nix should my package go?
+- Figuring out where an attribute is defined is a bit tricky:
+  - First one has to find the definition of it in all-packages.nix to see what file it refers to
+    - On GitHub this is even more problematic, as the `all-packages.nix` file is [too big to be displayed by GitHub](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/top-level/all-packages.nix)
+  - Then go to that file's definition, which takes quite some time for navigation (unless you have a plugin that can jump to it directly)
+    - It also slows down or even locks up editors due to the file size
+  - `nix edit -f . package-attr` works, though that's not yet stable (it relies on the `nix-command` feature being enabled) and doesn't work with packages that don't set `meta.position` correctly).
+- `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages
+
+# Detailed design
+[design]: #detailed-design
+
+This RFC establishes the standard of using `pkgs/unit/${shard}/${name}` "unit" directories for the definitions of the Nix packages `pkgs.${name}` in Nixpkgs, where `shard = toLower (substring 0 2 name)`.
+All unit directories are automatically discovered and incorporated into the `pkgs` set using `pkgs.${name} = pkgs.callPackage pkgs/unit/${shard}/${name}/pkg-fun.nix { }`.
+
+The following requirements will be checked by CI.
+This standard must be followed for newly added packages that can satisfy these requirements.
+A treewide migration to this standard will be performed for existing packages that can satisfy these requirements.
+
+## Structure
+
+The `pkgs/unit` directory must only contain unit directories, and only in subdirectories of the form `${shard}/${name}`.
+Each unit directory must contain at least a `pkg-fun.nix` file, but may contain arbitrary other files and directories.
+
+This ensures that maintainers don't have to verify this structure manually, which is prone to mistakes.
+
+## Only derivations
+
+If `pkgs/unit/${shard}/${name}` exists, `pkgs.${name}` must be a derivation that can be built directly with `nix-build`.
+
+This ensures that people can expect the unit directories to correspond to buildable packages and not functions like `pkgs.fetchFromGitHub` or `pkgs.buildRustCrate`.
+
+## Stable boundary
+
+Unit directories may only interact with the rest of Nixpkgs via the stable `pkgs.${name}` attributes, not with file references:
+- Files inside a unit directory must not reference files outside that unit directory.
```

Comment:
how does it work with update scripts that do e.g. `-I nixpkgs=../../../.`

Reply:
Yeah that's a good point, it was also mentioned in the Matrix channel. I should add a generic mention that such references should be ensured to not break. `CODEOWNERS` is another such case. This should be doable semi-manually, since it's not done that often (same for such update scripts).

Reply:
I'd say we should rather define a standard for how update script are called, which defines in which directory the script will be run and –if necessary– how to find the current nixpgks' path. Maybe write two sentences about this into the "Future work" section?

-----

For CODEOWNERS, I think this should be noted in the drawbacks section, because people/teams won't be able to subscribe to specific package ecosystems like Python or Gnome anymore.

Reply:
Replace with calls to

```
git rev-parse --show-toplevel

Reply:
@mweinelt: Doesn't work in shebangs like [here](https://github.com/NixOS/nixpkgs/blob/cb2d5a2fa9f2fa6dd2a619fc3be3e2de21a6a2f4/pkgs/applications/version-management/cz-cli/generate-dependencies.sh). I added a mention of this to the RFC though, will definitely have to make sure these don't break. I don't think it's in scope for this RFC to improve the update script situation, I'll add it to future work though.